README.md
# Timed Backend

[![Build Status](https://github.com/adfinis/timed-backend/workflows/Test/badge.svg)](https://github.com/adfinis/timed-backend/actions?query=workflow%3A%22Test%22)
[![Coverage](https://img.shields.io/badge/coverage-100%25-brightgreen.svg)](https://github.com/adfinis/timed-backend/blob/master/setup.cfg)
[![Pyup](https://pyup.io/repos/github/adfinis/timed-backend/shield.svg)](https://pyup.io/account/repos/github/adfinis/timed-backend/)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/adfinis/timed-backend)
[![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)

Timed timetracking software REST API built with Django

## Installation

**Requirements**

- docker
- docker-compose

After installing and configuring those requirements, you should be able to run the following
commands to complete the installation:

Add the `timed.local` entries to your hosts file:
```bash
echo "127.0.0.1 timed.local" | sudo tee -a /etc/hosts
```

Then just start the docker-compose setup:
```bash
make start
```

This brings up complete local installation, including our [Timed Frontend](https://github.com/adfinis/timed-frontend) project.

You can visit it at [http://timed.local](http://timed.local).

The API can be accessed at [http://timed.local/api/v1](http://timed.local/api/v1) and the admin interface at [http://timed.local/admin/](http://timed.local/admin/).

The Keycloak admin interface can be accessed at [http://timed.local/auth/admin](http://timed.local/auth/admin) with the account `admin` and password `admin`

## Development

To get the application working locally for development, make sure to create a file `.env` with the following content:

```
ENV=dev
DJANGO_OIDC_CREATE_USER=True
```

If you have existing users from the previous LDAP authentication, you want to add this line as well:

```
DJANGO_OIDC_USERNAME_CLAIM=preferred_username
```

The test data includes 3 users admin, fritzm and alexs with you can log into [http://timed.local](http://timed.local)

The username and password are identical.

To access the Django admin interface you will have to change the admin password in Django directly:

```console
$ make bash
root@0a036a10f3c4:/app# poetry run python manage.py changepassword admin
Changing password for user 'admin'
Password: 
Password (again): 
Password changed successfully for user 'admin'
```

Then you'll be able to login in the Django admin interface [http://timed.local/admin/](http://timed.local/admin/).


### Adding a user

If you want to add other users with different roles, add them in the Keycloak interface (as they would be coming from your LDAP directory).
You will also have to correct their employment in the Django admin interface as it is not correctly set for the moment.
Head to [http://timed.local/admin/](http://timed.local/admin/) after having perform a first login with the user.
You should see that new user in the `Employment -> Users`.
Click on the user and scroll down to the `Employments` section to set a `Location`.
Save the user and you should now see the _Timed_ interface correctly under that account.

## Configuration

Following options can be set as environment variables to configure Timed backend in documented [format](https://github.com/joke2k/django-environ#supported-types)
according to type.

| Parameter                                    | Description                                                                                           | Default                                                      |
|----------------------------------------------|-------------------------------------------------------------------------------------------------------|--------------------------------------------------------------|
| `DJANGO_ENV_FILE`                            | Path to setup environment vars in a file                                                              | .env                                                         |
| `DJANGO_DEBUG`                               | Boolean that turns on/off debug mode                                                                  | False                                                        |
| `DJANGO_SECRET_KEY`                          | Secret key for cryptographic signing                                                                  | not set (required)                                           |
| `DJANGO_ALLOWED_HOSTS`                       | List of hosts representing the host/domain names                                                      | not set (required)                                           |
| `DJANGO_HOST_PROTOCOL`                       | Protocol host is running on (http or https)                                                           | http                                                         |
| `DJANGO_HOST_DOMAIN`                         | Main host name server is reachable on                                                                 | not set (required)                                           |
| `DJANGO_DATABASE_NAME`                       | Database name                                                                                         | timed                                                        |
| `DJANGO_DATABASE_USER`                       | Database username                                                                                     | timed                                                        |
| `DJANGO_DATABASE_HOST`                       | Database hostname                                                                                     | localhost                                                    |
| `DJANGO_DATABASE_PORT`                       | Database port                                                                                         | 5432                                                         |
| `DJANGO_OIDC_DEFAULT_BASE_URL`               | Base URL of the OIDC provider                                                                         | http://timed.local/auth/realms/timed/protocol/openid-connect |
| `DJANGO_OIDC_OP_AUTHORIZATION_ENDPOINT`      | OIDC /auth endpoint                                                                                   | {`DJANGO_OIDC_DEFAULT_BASE_URL`}/auth                        |
| `DJANGO_OIDC_OP_TOKEN_ENDPOINT`              | OIDC /token endpoint                                                                                  | {`DJANGO_OIDC_DEFAULT_BASE_URL`}/token                       |
| `DJANGO_OIDC_OP_USER_ENDPOINT`               | OIDC /userinfo endpoint                                                                               | {`DJANGO_OIDC_DEFAULT_BASE_URL`}/userinfo                    |
| `DJANGO_OIDC_OP_JWKS_ENDPOINT`               | OIDC /certs endpoint                                                                                  | {`DJANGO_OIDC_DEFAULT_BASE_URL`}/certs                       |
| `DJANGO_OIDC_RP_CLIENT_ID`                   | Client ID by your OIDC provider                                                                       | timed-public                                                 |
| `DJANGO_OIDC_RP_CLIENT_SECRET`               | Client secret by your OIDC provider, should be None (flow start is handled by frontend)               | not set                                                      |
| `DJANGO_OIDC_RP_SIGN_ALGO`                   | Algorithm the OIDC provider uses to sign ID tokens                                                    | RS256                                                        |
| `DJANGO_OIDC_VERIFY_SSL`                     | Verify SSL on OIDC request                                                                            | dev: False, prod: True                                       |
| `DJANGO_OIDC_CREATE_USER`                    | Create new user if it doesn't exist in the database                                                   | False                                                        |
| `DJANGO_OIDC_USERNAME_CLAIM`                 | Username token claim for user lookup / creation                                                       | sub                                                          |
| `DJANGO_OIDC_EMAIL_CLAIM`                    | Email token claim for creating new users (if `DJANGO_OIDC_CREATE_USER` is enabled)                    | email                                                        |
| `DJANGO_OIDC_FIRSTNAME_CLAIM`                | First name token claim for creating new users (if `DJANGO_OIDC_CREATE_USER` is enabled)               | given_name                                                   |
| `DJANGO_OIDC_LASTNAME_CLAIM`                 | Last name token claim for creating new users (if `DJANGO_OIDC_CREATE_USER` is enabled)                | family_name                                                  |
| `DJANGO_OIDC_BEARER_TOKEN_REVALIDATION_TIME` | Time (in seconds) to cache a bearer token before revalidation is needed                               | 60                                                           |
| `DJANGO_OIDC_CHECK_INTROSPECT`               | Use token introspection for confidential clients                                                      | True                                                         |
| `DJANGO_OIDC_OP_INTROSPECT_ENDPOINT`         | OIDC token introspection endpoint (if `DJANGO_OIDC_CHECK_INTROSPECT` is enabled)                      | {`DJANGO_OIDC_DEFAULT_BASE_URL`}/token/introspect            |
| `DJANGO_OIDC_RP_INTROSPECT_CLIENT_ID`        | OIDC client id (if `DJANGO_OIDC_CHECK_INTROSPECT` is enabled) of confidential client                  | timed-confidential                                           |
| `DJANGO_OIDC_RP_INTROSPECT_CLIENT_SECRET`    | OIDC client secret (if `DJANGO_OIDC_CHECK_INTROSPECT` is enabled) of confidential client              | not set                                                      |
| `DJANGO_OIDC_ADMIN_LOGIN_REDIRECT_URL`       | URL of the django-admin, to which the user is redirected after successful admin login                 | dev: http://timed.local/admin/, prod: not set                |
| `DJANGO_ALLOW_LOCAL_LOGIN`                   | Enable / Disable login with local user/password (in admin)                                            | True                                                         |
| `EMAIL_URL`                                  | Uri of email server                                                                                   | smtp://localhost:25                                          |
| `DJANGO_DEFAULT_FROM_EMAIL`                  | Default email address to use for various responses                                                    | webmaster@localhost                                          |
| `DJANGO_SERVER_EMAIL`                        | Email address error messages are sent from                                                            | root@localhost                                               |
| `DJANGO_ADMINS`                              | List of people who get error notifications                                                            | not set                                                      |
| `DJANGO_WORK_REPORT_PATH`                    | Path of custom work report template                                                                   | not set                                                      |
| `DJANGO_SENTRY_DSN`                          | Sentry DSN for error reporting                                                                        | not set, set to enable Sentry integration                    |
| `DJANGO_SENTRY_TRACES_SAMPLE_RATE`           | Sentry trace sample rate, Set 1.0 to capture 100% of transactions                                     | 1.0                                                          |
| `DJANGO_SENTRY_SEND_DEFAULT_PII`             | Associate users to errors in Sentry                                                                   | True                                                         |
| `HURRICANE_REQ_QUEUE_LEN`                    | Django Hurricane's request queue length. When full, the readiness probe toggles                       | 250                                                          |
| `STATIC_ROOT`                                | Path to the static files. In prod, you may want to mount a docker volume here, so it can be served by nginx | `/app/static`                                                |
| `STATIC_URL`                                 | URL path to the static files on the web server. Configure nginx to point this to `$STATIC_ROOT`       | `/static`                                                    |

## Contributing

Look at our [contributing guidelines](CONTRIBUTING.md) to start with your first contribution.

## License

Code released under the [GNU Affero General Public License v3.0](LICENSE).


CHANGELOG.md
# v3.0.7

### Fix
* **redmine:** Convert Decimal objects to floats ([`5f613d1`](https://github.com/adfinis/timed-backend/commit/5f613d1e83f007270228c109846fb6a525eced71))

# v3.0.6

### Feature
* **redmine:** Add redmine issue id to log ([`b5c509b`](https://github.com/adfinis/timed-backend/commit/b5c509bc196137b28e84b814d4dba0a60b117747))

# v3.0.5

### Fix
* **tracking:** Fix updating own rejected reports and rejecting own reports ([`6a5d0ed`](https://github.com/adfinis/timed-backend/commit/6a5d0eda470939c59ad9ea869d4296e0115dd33e))

# v3.0.4

### Fix
* **redmine:** Log estimated_hours in update_project_expenditure command ([`fc1f631`](https://github.com/adfinis/timed-backend/commit/fc1f631a7fcdc25ab93b5cbcf38845f30af3f4a5))

# v3.0.3

### Fix
* **redmine:** Fix value check for custom fields ([`5f6bc53`](https://github.com/adfinis/timed-backend/commit/5f6bc532c6e0d76c9dae07b423afa6ea7c2ab52c))

# v3.0.2

### Fix
* **redmine:** Fix NoneType for amount offered/invoiced for projects ([`6e1f4c8`](https://github.com/adfinis/timed-backend/commit/6e1f4c89672a991ce1765bbed7c5e71f02a119e2))

# v3.0.1

### Feature
* Empty sums in correcr ordering ([`757de4e`](https://github.com/adfinis/timed-backend/commit/757de4e263cd42bb8521bccdd51dc6bf2207e761))
* **statistics:** Support ordering in new queryset wrapper ([`fb5a2dc`](https://github.com/adfinis/timed-backend/commit/fb5a2dc6480936004d90c067161905196aad58e0))

### Fix
* **tracking:** Fix report update notifactions ([`8d0d0fd`](https://github.com/adfinis/timed-backend/commit/8d0d0fd62896652a15ed00b840090cccdc4eaac8))
* **tests:** Customer statistic test had a missing customer ([`c99b512`](https://github.com/adfinis/timed-backend/commit/c99b5120fce4f3dc8bddf21346eb46bb8ba72239))
* **pytest:** Ignore "invalid escape sequence" deprecation warning ([`4e08672`](https://github.com/adfinis/timed-backend/commit/4e086727a1e69e658e5c044930ce1248aa9c1435))
* **statistics:** Refactor multiqs to use filtering aggregates ([`345b8df`](https://github.com/adfinis/timed-backend/commit/345b8df559593b03f69dcc1b81c590a4277d8fda))
* **makefile:** Use aliases for debug backend ([`9c47123`](https://github.com/adfinis/timed-backend/commit/9c47123af4cab3ab2095b9ff1b0e63ca973ee6ac))

# v3.0.0

### Feature
* **filters:** Add number multi value filter ([`4688e41`](https://github.com/adfinis/timed-backend/commit/4688e41da5300d789ed50bdd6af34d7d481767c8))
* **redmine:** Add pretend mode to redmine commands ([`abc5083`](https://github.com/adfinis/timed-backend/commit/abc50834feb3f84d3018abaa31073ab68e79dd76))
* **notifications:** Project budget check notifications ([`b81e28e`](https://github.com/adfinis/timed-backend/commit/b81e28e9d0b8386e54caf57b90960e392d5811c0))
* **statistics:** Show amount offered and invoiced in project statistics ([`144444b`](https://github.com/adfinis/timed-backend/commit/144444b298f2139f44a8ca291ca34ccfb3f66899))
* **redmine:** Import project expenditure from redmine ([`766f79b`](https://github.com/adfinis/timed-backend/commit/766f79bc17ce927a05217e9bacc18c478404e6f6))
* **redmine:** Update expenditures on redmine projects ([`0aa9da6`](https://github.com/adfinis/timed-backend/commit/0aa9da69e8432a4d9537b65dd935bebe23fd4c72))
* **filters:** Allow filtering of tasks and reports in statistics ([`b5b9c8d`](https://github.com/adfinis/timed-backend/commit/b5b9c8d633a4d6fc8633594349feec7ee59fb8d0))
* **employment:** Add is_external filter for user endpoint ([`8a1b272`](https://github.com/adfinis/timed-backend/commit/8a1b2723147c9775cb06abd090ec307610a2d254))
* **admin:** Add searchable dropdowns for user lists in admin ([`4c01054`](https://github.com/adfinis/timed-backend/commit/4c010542ce3b3544c3e87f1dd2ca7ff8ec4df245))
* Track remaining effort on tasks ([`3d045f2`](https://github.com/adfinis/timed-backend/commit/3d045f21ed7fd2147b49c6190dc3e1474c69decb))
* **tracking:** Reject reports ([`a4e8983`](https://github.com/adfinis/timed-backend/commit/a4e8983265d0b87101a6151982fbb8a802e4cd9a))

### Fix
* **tracking:** Fix automatic unreject when bulk updating ([`f110eb0`](https://github.com/adfinis/timed-backend/commit/f110eb0ea864a7115f7ed1d24e868aafb6c038f2))
* **tracking:** Fix remaining effort on report creation ([`abceb32`](https://github.com/adfinis/timed-backend/commit/abceb322e042df5c34b04e685c331527848c898f))
* **tracking:** Fix setting of remaining effort ([`16f1dbb`](https://github.com/adfinis/timed-backend/commit/16f1dbb54f625a8468fd33066a685ac1cfae7fec))
* **notifications:** Omit projects with no reports ([`91a6dd5`](https://github.com/adfinis/timed-backend/commit/91a6dd5ec2d128d6df3994c78eebbb295ec9a2f5))
* **tracking:** Allow null values on remaining effort for reports ([`08a5aa4`](https://github.com/adfinis/timed-backend/commit/08a5aa429eac6d25cec0699a42919ee8f959ed12))
* **tracking:** Fix  absence for users with multiple employments ([`d884ef6`](https://github.com/adfinis/timed-backend/commit/d884ef6a4463e5095fcecc6cd999aa6b595f5530))
* Add missing rejected field to ReportIntersectionSerializer ([`ee8f79a`](https://github.com/adfinis/timed-backend/commit/ee8f79a1a724763bdd51222010f47ec40ef71622))
* **auth:** Let failing auth requests return 401 ([`8454601`](https://github.com/adfinis/timed-backend/commit/8454601019f33272a39814ac8e3fe033c758e7e7))
* **dev:** Remove deprecated flag from pre-commit isort ([`50e5da2`](https://github.com/adfinis/timed-backend/commit/50e5da2ad5ef12098e0128ba907ac40ac2fa1773))
* **tracking:** Fix remaining effort check when creating report ([`fc7c92c`](https://github.com/adfinis/timed-backend/commit/fc7c92cf0f3cb937100616abb24bd06804408a51))
* **statistics:** Add missing fields for project and task statistics ([`89fb718`](https://github.com/adfinis/timed-backend/commit/89fb718901f41914323a60d99a2983ba0454daa0))
* **reports:** Fix project and customer statistics ([`a3ab8ac`](https://github.com/adfinis/timed-backend/commit/a3ab8acb5be4107ea7f4f6677cdbdb57dd0b95c2))

timed/settings.py
import os
import re

import environ
import sentry_sdk
from pkg_resources import resource_filename
from sentry_sdk.integrations.django import DjangoIntegration

env = environ.Env()

django_root = environ.Path(__file__) - 2

ENV_FILE = env.str("DJANGO_ENV_FILE", default=django_root(".env"))
if os.path.exists(ENV_FILE):  # pragma: no cover
    environ.Env.read_env(ENV_FILE)

# per default production is enabled for security reasons
# for development create .env file with ENV=dev
ENV = env.str("ENV", "prod")


def default(default_dev=env.NOTSET, default_prod=env.NOTSET):
    """Environment aware default."""
    return default_prod if ENV == "prod" else default_dev


# Database definition

DATABASES = {
    "default": {
        "ENGINE": env.str(
            "DJANGO_DATABASE_ENGINE", default="django_prometheus.db.backends.postgresql"
        ),
        "NAME": env.str("DJANGO_DATABASE_NAME", default="timed"),
        "USER": env.str("DJANGO_DATABASE_USER", default="timed"),
        "PASSWORD": env.str("DJANGO_DATABASE_PASSWORD", default=default("timed")),
        "HOST": env.str("DJANGO_DATABASE_HOST", default="localhost"),
        "PORT": env.str("DJANGO_DATABASE_PORT", default=""),
    }
}


# Application definition

DEBUG = env.bool("DJANGO_DEBUG", default=default(True, False))
SECRET_KEY = env.str("DJANGO_SECRET_KEY", default=default("uuuuuuuuuu"))
ALLOWED_HOSTS = env.list("DJANGO_ALLOWED_HOSTS", default=default(["*"]))
HOST_PROTOCOL = env.str("DJANGO_HOST_PROTOCOL", default=default("http"))
HOST_DOMAIN = env.str("DJANGO_HOST_DOMAIN", default=default("localhost:4200"))


INSTALLED_APPS = [
    "timed.apps.TimedAdminConfig",
    "django.contrib.humanize",
    "multiselectfield.apps.MultiSelectFieldConfig",
    "django.forms",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "rest_framework",
    "django_filters",
    "djmoney.apps.MoneyConfig",
    "mozilla_django_oidc",
    "django_prometheus.apps.DjangoPrometheusConfig",
    "corsheaders",
    "nested_inline",
    "timed.employment",
    "timed.projects",
    "timed.tracking",
    "timed.reports",
    "timed.redmine",
    "timed.subscription",
    "timed.notifications",
]

if ENV == "dev":
    INSTALLED_APPS.append("django_extensions")

MIDDLEWARE = [
    "django_prometheus.middleware.PrometheusBeforeMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "django_prometheus.middleware.PrometheusAfterMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
]

ROOT_URLCONF = "timed.urls"

FORM_RENDERER = "django.forms.renderers.TemplatesSetting"
TEMPLATES = [
    # default: needed for django-admin
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [django_root("timed", "templates")],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ]
        },
    },
    # template backend for plain text (no escaping)
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "NAME": "text",
        "APP_DIRS": True,
        "OPTIONS": {
            "autoescape": False,
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "timed.wsgi.application"


# Internationalization
# https://docs.djangoproject.com/en/1.9/topics/i18n/

LOCALE_PATHS = [django_root("timed", "locale")]

LANGUAGE_CODE = "en-US"

TIME_ZONE = env.str("DJANGO_TIME_ZONE", "Europe/Zurich")

USE_I18N = True
USE_L10N = True

DATETIME_FORMAT = env.str("DJANGO_DATETIME_FORMAT", "d.m.Y H:i:s")
DATE_FORMAT = env.str("DJANGO_DATE_FORMAT", "d.m.Y")
TIME_FORMAT = env.str("DJANGO_TIME_FORMAT", "H:i:s")

DECIMAL_SEPARATOR = env.str("DECIMAL_SEPARATOR", ".")

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.9/howto/static-files/

STATIC_URL = env.str("STATIC_URL", "/static/")
STATIC_ROOT = env.str("STATIC_ROOT", "/app/static")

# Cache

CACHES = {
    "default": {
        "BACKEND": env.str(
            "CACHE_BACKEND",
            default="django_prometheus.cache.backends.locmem.LocMemCache",
        ),
        "LOCATION": env.str("CACHE_LOCATION", ""),
    }
}

# Rest framework definition

REST_FRAMEWORK = {
    "DEFAULT_FILTER_BACKENDS": (
        "django_filters.rest_framework.DjangoFilterBackend",
        "rest_framework.filters.SearchFilter",
        "rest_framework.filters.OrderingFilter",
    ),
    "DEFAULT_PARSER_CLASSES": ("rest_framework_json_api.parsers.JSONParser",),
    "DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.IsAuthenticated",),
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "mozilla_django_oidc.contrib.drf.OIDCAuthentication",
    ),
    "DEFAULT_METADATA_CLASS": "rest_framework_json_api.metadata.JSONAPIMetadata",
    "EXCEPTION_HANDLER": "rest_framework_json_api.exceptions.exception_handler",
    "DEFAULT_PAGINATION_CLASS": "rest_framework_json_api.pagination.JsonApiPageNumberPagination",
    "DEFAULT_RENDERER_CLASSES": ("rest_framework_json_api.renderers.JSONRenderer",),
    "TEST_REQUEST_RENDERER_CLASSES": (
        "rest_framework_json_api.renderers.JSONRenderer",
        "rest_framework.renderers.JSONRenderer",
        "rest_framework.renderers.MultiPartRenderer",
    ),
    "TEST_REQUEST_DEFAULT_FORMAT": "vnd.api+json",
}

JSON_API_FORMAT_FIELD_NAMES = "dasherize"
JSON_API_FORMAT_TYPES = "dasherize"
JSON_API_PLURALIZE_TYPES = True

APPEND_SLASH = False

# Authentication

AUTH_USER_MODEL = "employment.User"
AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
    "timed.authentication.TimedOIDCAuthenticationBackend",
]

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"  # noqa
    },
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},  # noqa
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},  # noqa
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"  # noqa
    },
]

# OIDC

OIDC_DEFAULT_BASE_URL = env.str(
    "DJANGO_OIDC_DEFAULT_BASE_URL",
    default="http://timed.local/auth/realms/timed/protocol/openid-connect",
)
OIDC_OP_AUTHORIZATION_ENDPOINT = env.str(
    "DJANGO_OIDC_OP_AUTHORIZATION_ENDPOINT", default=f"{OIDC_DEFAULT_BASE_URL}/auth"
)

OIDC_OP_TOKEN_ENDPOINT = env.str(
    "DJANGO_OIDC_OP_TOKEN_ENDPOINT", default=f"{OIDC_DEFAULT_BASE_URL}/token"
)
OIDC_OP_USER_ENDPOINT = env.str(
    "DJANGO_OIDC_OP_USER_ENDPOINT", default=f"{OIDC_DEFAULT_BASE_URL}/userinfo"
)
OIDC_OP_JWKS_ENDPOINT = env.str(
    "DJANGO_OIDC_OP_JWKS_ENDPOINT", default=f"{OIDC_DEFAULT_BASE_URL}/certs"
)

OIDC_RP_CLIENT_ID = env.str("DJANGO_OIDC_RP_CLIENT_ID", default="timed-public")
OIDC_RP_CLIENT_SECRET = env.str("DJANGO_OIDC_RP_CLIENT_SECRET", default=None)

OIDC_VERIFY_SSL = env.bool("DJANGO_OIDC_VERIFY_SSL", default=default(False, True))
OIDC_RP_SIGN_ALGO = env.str("DJANGO_OIDC_RP_SIGN_ALGO", default="RS256")

OIDC_CREATE_USER = env.bool("DJANGO_OIDC_CREATE_USER", default=True)
OIDC_USERNAME_CLAIM = env.str(
    "DJANGO_OIDC_USERNAME_CLAIM", default=default("preferred_username", "sub")
)
OIDC_EMAIL_CLAIM = env.str("DJANGO_OIDC_EMAIL_CLAIM", default="email")
OIDC_FIRSTNAME_CLAIM = env.str("DJANGO_OIDC_FIRSTNAME_CLAIM", default="given_name")
OIDC_LASTNAME_CLAIM = env.str("DJANGO_OIDC_LASTNAME_CLAIM", default="family_name")
# time in seconds
OIDC_BEARER_TOKEN_REVALIDATION_TIME = env.int(
    "DJANGO_OIDC_BEARER_TOKEN_REVALIDATION_TIME", default=60
)

# introspection endpoint for checking confidential client authentication
OIDC_CHECK_INTROSPECT = env.bool("DJANGO_OIDC_CHECK_INTROSPECT", default=True)
OIDC_OP_INTROSPECT_ENDPOINT = env.str(
    "DJANGO_OIDC_OP_INTROSPECT_ENDPOINT",
    default=f"{OIDC_DEFAULT_BASE_URL}/token/introspect",
)
OIDC_RP_INTROSPECT_CLIENT_ID = env.str(
    "DJANGO_OIDC_RP_INTROSPECT_CLIENT_ID", default="timed-confidential"
)
OIDC_RP_INTROSPECT_CLIENT_SECRET = env.str(
    "DJANGO_OIDC_RP_INTROSPECT_CLIENT_SECRET", default=None
)

# admin page after completing server-side authentication flow
LOGIN_REDIRECT_URL = env.str(
    "DJANGO_OIDC_ADMIN_LOGIN_REDIRECT_URL", default=default("http://timed.local/admin/")
)

# allow / disallow login with local user / password
ALLOW_LOCAL_LOGIN = env.bool("DJANGO_ALLOW_LOCAL_LOGIN", default=True)

if not ALLOW_LOCAL_LOGIN:  # pragma: no cover
    APPLICATION_BACKENDS = ["timed.authentication.TimedOIDCAuthenticationBackend"]

# Email definition

EMAIL_CONFIG = env.email_url("EMAIL_URL", default="smtp://localhost:25")
vars().update(EMAIL_CONFIG)

EMAIL_USE_TLS = env.bool("DJANGO_EMAIL_USE_TLS", default=True)

DEFAULT_FROM_EMAIL = env.str(
    "DJANGO_DEFAULT_FROM_EMAIL", default("webmaster@localhost")
)

CUSTOMER_CENTER_EMAIL = env.str(
    "DJANGO_CUSTOMER_CENTER_EMAIL", default("admin@localhost")
)

SERVER_EMAIL = env.str("DJANGO_SERVER_EMAIL", default("root@localhost"))
EMAIL_EXTRA_HEADERS = {"Auto-Submitted": "auto-generated"}


def parse_admins(admins):
    """
    Parse env admins to django admins.

    Example of DJANGO_ADMINS environment variable:
    Test Example <test@example.com>,Test2 <test2@example.com>
    """
    result = []
    for admin in admins:
        match = re.search(r"(.+) \<(.+@.+)\>", admin)
        if not match:
            raise environ.ImproperlyConfigured(
                'In DJANGO_ADMINS admin "{0}" is not in correct '
                '"Firstname Lastname <email@example.com>"'.format(admin)
            )
        result.append((match.group(1), match.group(2)))
    return result


ADMINS = parse_admins(env.list("DJANGO_ADMINS", default=[]))


# Redmine definition (optional)

REDMINE_URL = env.str("DJANGO_REDMINE_URL", default="")
REDMINE_APIKEY = env.str("DJANGO_REDMINE_APIKEY", default="")
REDMINE_HTACCESS_USER = env.str("DJANGO_REDMINE_HTACCESS_USER", default="")
REDMINE_HTACCESS_PASSWORD = env.str("DJANGO_REDMINE_HTACCESS_PASSWORD", default="")
REDMINE_SPENTHOURS_FIELD = env.int("DJANGO_REDMINE_SPENTHOURS_FIELD", default=0)
REDMINE_AMOUNT_OFFERED_FIELD = env.int("DJANGO_REDMINE_AMOUNT_OFFERED_FIELD", default=1)
REDMINE_AMOUNT_INVOICED_FIELD = env.int(
    "DJANGO_REDMINE_AMOUNT_INVOICED_FIELD", default=2
)
REDMINE_BUILD_PROJECT = env.str("DJANGO_REDMINE_BUILD_PROJECT", default="build")


# Work report definition

WORK_REPORT_PATH = env.str(
    "DJANGO_WORK_REPORT_PATH",
    default=resource_filename("timed.reports", "templates/workreport.ots"),
)

WORK_REPORTS_EXPORT_MAX_COUNT = env.int(
    "DJANGO_WORK_REPORTS_EXPORT_MAX_COUNT", default=0
)

REPORTS_EXPORT_MAX_COUNT = env.int("DJANGO_REPORTS_EXPORT_MAX_COUNT", default=0)

# Tracking: Report fields which should be included in email (when report was
# changed during verification)
TRACKING_REPORT_VERIFIED_CHANGES = env.list(
    "DJANGO_TRACKING_REPORT_VERIFIED_CHANGES",
    default=["task", "comment", "not_billable"],
)

# Sentry error tracking
if env.str("DJANGO_SENTRY_DSN", default=""):  # pragma: no cover
    sentry_sdk.init(
        dsn=env.str("DJANGO_SENTRY_DSN", default=""),
        integrations=[DjangoIntegration()],
        # Set traces_sample_rate to 1.0 to capture 100%
        # of transactions for performance monitoring.
        traces_sample_rate=env.float("DJANGO_SENTRY_TRACES_SAMPLE_RATE", default=1.0),
        # If you wish to associate users to errors (assuming you are using
        # django.contrib.auth) you may enable sending PII data.
        send_default_pii=env.bool("DJANGO_SENTRY_SEND_DEFAULT_PII", default=True),
        # By default the SDK will try to use the SENTRY_RELEASE
        # environment variable, or infer a git commit
        # SHA as release, however you may want to set
        # something more human-readable.
        # release="myapp@1.0.0",
    )

SECURE_PROXY_SSL_HEADER = ("HTTP_X_FORWARDED_PROTO", "https")
DATA_UPLOAD_MAX_NUMBER_FIELDS = env.int(
    "DJANGO_DATA_UPLOAD_MAX_NUMBER_FIELDS", default=1000
)
CORS_ALLOWED_ORIGINS = env.list("DJANGO_CORS_ALLOWED_ORIGINS", default=[])
DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

BUILD_PROJECTS = env.str("DJANGO_BUILD_PROJECT", default="_BUILD")
